va/va.go

```diff
@@ -199,17 +215,117 @@ func (va VAImpl) performValidation(task *vaTask, results chan<- *core.Validation
 	va.log.Printf("Sleeping for %s seconds before validating", time.Second*len)
 	va.clk.Sleep(time.Second * len)
 
-	// TODO(@cpu): Implement validation for DNS-01
 	switch task.Challenge.Type {
 	case acme.ChallengeHTTP01:
 		results <- va.validateHTTP01(task)
 	case acme.ChallengeTLSSNI02:
 		results <- va.validateTLSSNI02(task)
+	case acme.ChallengeDNS01:
+		results <- va.validateDNS01(task)
 	default:
 		va.log.Printf("Error: performValidation(): Invalid challenge type: %q", task.Challenge.Type)
 	}
 }
 
+func (va VAImpl) validateDNS01(task *vaTask) *core.ValidationRecord {
+	const dns01Prefix = "_acme-challenge"
+	challengeSubdomain := fmt.Sprintf("%s.%s", dns01Prefix, task.Identifier)
+
+	result := &core.ValidationRecord{
+		URL:         challengeSubdomain,
+		ValidatedAt: va.clk.Now(),
+	}
+
+	txts, err := va.lookupTXT(challengeSubdomain)
+	if err != nil {
+		result.Error = acme.UnauthorizedProblem("Error retrieving TXT records for DNS challenge")
+		return result
+	}
+
+	if len(txts) == 0 {
+		msg := fmt.Sprintf("No TXT records found for DNS challenge")
+		result.Error = acme.UnauthorizedProblem(msg)
+		return result
+	}
+
+	h := sha256.New()
+	task.Challenge.RLock()
+	h.Write([]byte(task.Challenge.KeyAuthorization))
+	task.Challenge.RUnlock()
+	authorizedKeysDigest := base64.RawURLEncoding.EncodeToString(h.Sum(nil))
+
+	for _, element := range txts {
+		if subtle.ConstantTimeCompare([]byte(element), []byte(authorizedKeysDigest)) == 1 {
+			return result
+		}
+	}
+
+	msg := fmt.Sprintf("Correct value not found for DNS challenge")
+	result.Error = acme.UnauthorizedProblem(msg)
+	return result
+}
+
+func (va VAImpl) lookupTXT(hostname string) ([]string, error) {
+	const qtype = dns.TypeTXT
+	var txt []string
+
+	dnsTimeout := time.Second * 5
+	ctx, _ := context.WithDeadline(context.Background(), va.clk.Now().Add(dnsTimeout))
+
+	m := new(dns.Msg)
+	m.SetQuestion(dns.Fqdn(hostname), qtype)
+	// Set DNSSEC OK bit for resolver
```

Comment:
This is not accurate. The docs for this:

> func (dns *Msg) SetEdns0(udpsize uint16, do bool) *Msg
>    SetEdns0 appends a EDNS0 OPT RR to the message. TSIG should always the
>    last RR in a message.

So this should be something like "accept large UDP replies." Also note that (a) we don't expect large replies because we don't evaluate DNS ourselves, and (b) if you're talking to 8.8.8.8, you're going over the public internet, which I think has a lower MTU than in-datacenter. So you'd expect IP fragmentation with a 4096-bit reply. I think let's delete this for now, and we can experiment with different settings if it causes real-world problems.